flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java

```diff
@@ -122,6 +117,9 @@ protected void finishTask() throws Exception {
     public Future<Boolean> triggerCheckpointAsync(
             CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions) {
         if (!isExternallyInducedSource) {
+            if (checkpointOptions.getCheckpointType().shouldDrain()) {
+                mainMailboxExecutor.execute(this::endData, "Drain pipeline on stop-with-savepoint");
+            }
```

Comment:
We need a test case what should happen if `stop-with-savepoint --drain` fails - job should failover, because we have already emitted `MAX_WATERMARK`/`endInput`